paddle/fluid/pir/transforms/onednn/matmul_reshape_add_pass.cc

```diff
@@ -0,0 +1,155 @@
+    paddle::drr::SourcePattern pat = ctx->SourcePattern();
+
+    const auto &matmul = pat.Op(paddle::dialect::MatmulOp::name(),
+                                {{"transpose_x", pat.Attr("trans_x")},
+                                 {"transpose_y", pat.Attr("trans_y")}});
+    pat.Tensor("matmul_out") = matmul(pat.Tensor("x"), pat.Tensor("w"));
+
+    const auto &full_int_array3 =
+        pat.Op(paddle::dialect::FullIntArrayOp::name());
+    pat.Tensor("shape") = full_int_array3();
+    const auto &reshape3 = pat.Op(paddle::dialect::ReshapeOp::name());
+    reshape3({&pat.Tensor("matmul_out"), &pat.Tensor("shape")},
+             {&pat.Tensor("out"), &pat.Tensor("xshape")});
+
+    const auto &add_ = pat.Op(paddle::dialect::AddOp::name());
+    pat.Tensor("add_out") = as_x_ ? add_(pat.Tensor("out"), pat.Tensor("y"))
+                                  : add_(pat.Tensor("y"), pat.Tensor("out"));
+
+    pat.AddConstraint([&](const paddle::drr::MatchContext &match_ctx) {
+      auto w_dtype = pir::GetDataTypeFromValue(match_ctx.Tensor("w"));
+      if (!w_dtype.isa<pir::Float16Type>() &&
+          !w_dtype.isa<pir::Float32Type>() &&
+          !w_dtype.isa<pir::Float64Type>()) {
+        return false;
+      }
+
+      auto w_dims = pir::GetShapeFromValue(match_ctx.Tensor("w"));
+      auto x_dims = pir::GetShapeFromValue(match_ctx.Tensor("x"));
+      auto y_dims = pir::GetShapeFromValue(match_ctx.Tensor("y"));
+      if (w_dims.size() != 2 || x_dims.size() < 2) {
+        return false;
+      }
+      // Currently，FcOp and GemmEpilogueOp support only RRR format
+      if (x_dims.at(x_dims.size() - 1) != w_dims.at(0) ||
+          match_ctx.Attr<bool>("trans_x") == true ||
+          match_ctx.Attr<bool>("trans_y") == true) {
+        return false;
+      }
+
+      if (y_dims.size() == 1) {
+        return y_dims.at(0) == w_dims.at(1);
+      }
+
+      if (y_dims.size() == 2) {
+        return y_dims.at(0) == 1 && y_dims.at(1) == w_dims.at(1);
+      }
+
+      return false;
+    });
+
+    paddle::drr::ResultPattern res = pat.ResultPattern();
+    const auto &in_num_col_dims_attr =
+        res.ComputeAttr([](const paddle::drr::MatchContext &match_ctx) -> int {
+          auto x_dims = pir::GetShapeFromValue(match_ctx.Tensor("x"));
+          return static_cast<int>(x_dims.size()) - 1;
+        });
+
+    const auto &fc_op = res.Op(fused_matmul_name_,
+                               {{
+                                   {"in_num_col_dims", in_num_col_dims_attr},
+                                   {"activation_type", res.StrAttr("")},
+                                   {"padding_weights", res.BoolAttr(false)},
+                               }});
+    fc_op({&res.Tensor("x"), &res.Tensor("w"), &res.Tensor("y")},
+          {&res.Tensor("matmul_out")});
+    const auto &reshape = res.Op(paddle::dialect::ReshapeOp::name());
+    reshape({&res.Tensor("matmul_out"), &res.Tensor("shape")},
```

Comment:
So this pattern will eventually fuse `matmul+reshpe+add` to `fc+reshape`? It's as expected? Cause this fusion might affect the later pass `fc_activation_fuse_pass`.